pkg/tbtc/node.go

```diff
@@ -205,3 +265,203 @@ func (n *node) joinDKGIfEligible(seed *big.Int, startBlockNumber uint64) {
 		logger.Infof("not eligible for DKG with seed [0x%x]", seed)
 	}
 }
+
+// dkgRetryLoop is a struct that encapsulates the DKG retry logic.
+type dkgRetryLoop struct {
+	memberIndex          group.MemberIndex
+	selectedOperators    chain.Addresses
+	inactiveOperatorsSet map[chain.Address]bool
+
+	chainConfig *ChainConfig
+
+	attemptCounter    uint
+	attemptStartBlock uint64
+
+	randomRetryCounter uint
+	randomRetrySeed    int64
+
+	delayBlocks              uint64
+	delayBlocksBumpFrequency uint
+	delayBlocksBumpFactor    uint64
+}
+
+func newDkgRetryLoop(
+	seed *big.Int,
+	initialStartBlock uint64,
+	memberIndex group.MemberIndex,
+	selectedOperators chain.Addresses,
+	chainConfig *ChainConfig,
+) *dkgRetryLoop {
+	// Pre-compute the 8-byte seed that may be needed for the random
+	// retry algorithm. Since the original DKG seed passed as parameter
+	// can have a variable length, it is safer to take the first 8 bytes
+	// of sha256(seed) as the randomRetrySeed.
+	seedSha256 := sha256.Sum256(seed.Bytes())
+	randomRetrySeed := int64(binary.BigEndian.Uint64(seedSha256[:8]))
+
+	return &dkgRetryLoop{
+		memberIndex:              memberIndex,
+		selectedOperators:        selectedOperators,
+		inactiveOperatorsSet:     make(map[chain.Address]bool),
+		chainConfig:              chainConfig,
+		attemptCounter:           0,
+		attemptStartBlock:        initialStartBlock,
+		randomRetryCounter:       0,
+		randomRetrySeed:          randomRetrySeed,
+		delayBlocks:              5,
+		delayBlocksBumpFrequency: 100,
+		delayBlocksBumpFactor:    20,
```

Comment:
This structure is currently a mix of constants like `delayBlocksBumpFactor` and mutable variables like `attemptCounter`.

I would find it easier to reason about if we separated out the constants somewhere else, that way we're calling something like

`delayBlocks *= DkgRetryConstants.DELAY_BLOCKS_BUMP_FACTOR` instead

Reply:
We may want to make them configurable in the near future. Those are not true constants.

Reply:
configurable on a by-loop level or by-client level?

Reply:
Those params must be the same for all group members so we cannot configure them by-client. However they can be the result of computations done using the same algorithm by all members. Unclear at the moment but I can imagine that we may need to shorten the delay if we are close to the timeout block.